pyccel/stdlib/lists/lists.c

```diff
@@ -0,0 +1,665 @@
+
+        Parameters
+        ----------
+        list : t_list*
+            the t_list to be reversed
+*/
+{
+    size_t tsize = tSizes[list->type];
+    size_t index = 0;
+    char *elements = (char*)(list->elements);
+    void *buffer = malloc(tsize);
+
+    while (index + 1 <= list->size / 2)
+    {
+        memcpy(buffer, &elements[index * tsize], tsize);
+        memcpy(&elements[index * tsize], &elements[(list->size - index - 1) * tsize], tsize);
+        memcpy(&elements[(list->size - index - 1) * tsize], buffer, tsize);
+        index++;
+    }
+    free(buffer);
+}
+
+
+void*   array_subscripting(t_list *list, size_t index)
+/*
+        provides access to a specific element in a t_list
+
+        Parameters
+        ----------
+        list : t_list*
+            the t_list of the element to be accessed
+
+        index : long int
+            the element to be accessed
+*/
+{
+    char *elements = (char*)(list->elements);
+
+    if (list->size > index)
+        return ((void*)(&elements[index * tSizes[list->type]]));
+    return (NULL);
+}
+
+// Sorting ////////////////////////////
+
+/**
+ * @brief Set the length and type object for group allocation.
+ * 
+ * @param list 
+ * @param len 
+ * @param type 
+ */
+void set_length_and_type(t_list *list, size_t *len, int8_t *type)
+{
+    t_list *tmp;
+
+    tmp = list;
+    *len = tmp->size;
+    while (tmp->type == lst_list)
+    {
+        tmp = ((t_list **)(tmp->elements))[0];
+        *len *= tmp->size;
+    }
+    *type = tmp->type;
+}
+
+/**
+ * @brief Branch recursively through every element in the list and copy 
+ * it's value in the array pointed to by `int8_t *group`.
+ * Returns the number of elements copied on every recursion.
+ * 
+ * @param list 
+ * @param group 
+ * @param offset 
+ * @return size_t 
+ */
+size_t collect_data(t_list *list, int8_t *group, size_t offset)
+{
+    if (list->type == lst_list)
+    {
+        for (size_t i = 0; i < list->size; i++)
+            offset += collect_data(((t_list **)list->elements)[i], group, offset);
+    }
+    else
+    {
+        memcpy(group + offset*tSizes[list->type], list->elements, list->size);
+        offset += list->size;
+    }
+    return (offset);
+}
+
+/**
+ * @brief Allocate and collect the sub-elements of a list.
+ * 
+ * @param list 
+ * @param len 
+ * @param type 
+ * @return int8_t* 
+ */
+int8_t *group_node_items(t_list *list, size_t len, int8_t type)
+{
+    int8_t *group;
+
+    group = (int8_t *)malloc(len * tSizes[type]);
+    collect_data(list, group, 0);
+    return (group);
+}
+
+/**
+ * @brief If the list consists of pointers over other lists, it attempts
+ * to collect the sub-elements of the lists pointed to by the indexs `i1`
+ * and `i2`, putting them in two seperate allocated arrays, then compare 
+ * every value.
+ * Returns comparision value.
+ * 
+ * @param list 
+ * @param i1 
+ * @param i2 
+ * @return int 
+ */
+int compare(t_list *list, int i1, int i2)
```

Comment:
What is this function comparing? I expected it to compare 2 lists but it seems that it compares 2 elements is that right? If so why is a function needed for this?

Reply:
Ok, the list could contain numerical values or be a list of lists. If it happens that the list consist of numerical values (rank=1), the function would simply compare the elements in the indexes `i1` and `i2` after casting to the corresponding type. But for list of lists (rank>1), deciding whether the element in the index `i1` is bigger or smaller than `i2` won't be possible unless we compare the values in the innermost dimension.

Reply:
But if you want to compare an element of a list of lists to another element of a list of lists then this is equivalent to comparing a list.

I expected to see something like:
```c
/************************************
*  Function to compare 2 lists
*  Returns 1 if list1>list2
*  Returns -1 if list2>list1
*  Returns 0 if list1==list2
************************************/
int compare(t_list* list1, t_list* list2)
{
    const int list_len = MIN(list1->size, list2->size);
    for(int i(0); i<list_len; ++i)
    {
        if (list1->type == lst_list)
        {
            // assuming list2->type == lst_list for valid comparison
            int const comp = compare(list1->elements[i], list2->elements[i]);
            if (comp != 0)
                return comp;
        }
        else if (list1->elements[i] < list2->elements[i])
        {
            return -1;
        }
        else if (list1->elements[i] > list2->elements[i])
        {
            return 1;
        }
    }
    if (list1->size < list2->size) {
        return -1;
    }
    else if (list1->size > list2->size) {
        return 1;
    }
    else {
        return 0;
    }
}
```
Then if you want to compare 2 elements you do that directly. Either by knowing that they are both literals and comparing via that, or knowing that they are both lists and calling this function

Reply:
This is odd, I don't understand why you want to compare the size of the lists. Haven't we agreed that the list will be homogeneous?

Reply:
Why would 2 unrelated homogeneous lists have the same size?

To be clear my function handles things like:
```
[1,2,3]<[1,2]
[1,2,3]<[1,2,4]
[5,3,1]<[7,32,1,9]
```
If you mean 2 elements of a list of list. I don't think homogeneous lists need to have this restriction. It is required for arrays because all the data is stored in the same memory block. This is not the case here

Reply:
https://github.com/pyccel/pyccel/pull/1192/commits/e7d7fce524d55ba8528fe125504613a3bdb7fb27